### PR TITLE
Add merge parameter to method functions

### LINKED ISSUE #103: Append output variables from functions to input dataset
Not every function will fit the `fn(ds: Dataset, ...) -> Dataset` signature,  but for the large majority that do we have so far been adopting a convention that only returns newly created variables.   Another option would be to always try to write those variables into the input dataset.  For lack of a better phrase I'll call the former "Ex situ" updates and the latter "In situ" updates.  Here are some pros/cons of each:

### Ex situ updates

- **Advantages**
  - It makes it easier to transform/interrogate results before deciding to merge them
  - Merging datasets in Xarray is trivial when there is no index but if result variables are indexed and the provided dataset is not, there is any argument to be made in leaving what to do up to the user
  - Function style and it's less code to write/maintain
- **Disadvantages**
  - `ds.merge(fn(ds))` is clunky if you wanted the variables merged in the first place
  - It's not as nice in pipelines, e.g. you have to do this instead: `ds = xr.merge([ fn1(ds), fn2(ds) ])`

### In situ updates

- **Advantages**
  - Pipelines are nicer, e.g. `ds.pipe(count_alleles).pipe(allele_frequency)`
- **Disadvantages**
  - We're likely to have some functions that produce 10s (or possibly 100s) of variables and users will only care about a few of them so fully merging all of them with a working dataset will almost never be the intended use (e.g. [nirvana]( https://hail.is/docs/0.2/methods/genetics.html#hail.methods.nirvana)/[vep](https://hail.is/docs/0.2/methods/genetics.html#hail.methods.vep))

My more opinionated view of this is that the ex situ updates are best in larger, more sophisticated workflows.  I say that because I think those workflows will be very much dominated by Xarray code with calls to individual sgkit functions interspersed within it, so the need for chaining sgkit operations is not very high and the need to transform/interrogate results before merging them in is higher.

It would be great to find some way to get all of those advantages above though.  Perhaps there is a way to do something like this more elegantly:

```python
# Assuming `fn` makes inplace updates
ds_only_new_vars = fn(ds)[fn.output_variables]
```

A potential issue with this is that functions may rely on variables from other functions (see https://github.com/pystatgen/sgkit/pull/102#issue-465519140) and add them when necessary, which would mean the set of output variables isn't always the same.  I don't think this is actually an issue with Dask as long as the definition of the variables is fast.  It would be for most functions and Dask will remove any redundancies, so presumably we wouldn't have to necessarily return or add them (i.e. this affects both the in situ and ex situ strategies).  Using numpy variables definitely wouldn't work that way but I'm not sure that use case is as practical outside of testing anyhow.

Some of @ravwojdyla's work to "schemify" function results would make it easy to know what variables a function produces (https://github.com/pystatgen/sgkit/issues/43#issuecomment-669478348).  There could be some overlap here with how that works out too.

### PR DESCRIPTION
**Pull request by tomwhite**
Fixes #103

### CODE REVIEW

File: sgkit/stats/association.py [hunk 1]
```diff
@@ -203,10 +211,11 @@ def gwas_linear_regression(
...
     res = linear_regression(G.T, X, Y)
-    return xr.Dataset(
+    new_ds = xr.Dataset(
         {
             "variant_beta": (("variants", "traits"), res.beta),
             "variant_t_value": (("variants", "traits"), res.t_value),
             "variant_p_value": (("variants", "traits"), res.p_value),
         }
     )
+    return merge_datasets(ds, new_ds) if merge else new_ds
```
Comment (jeromekelleher) [inline sgkit/stats/association.py:221]: I wonder if it's worth having a function that does the conditional merge? I can see it being a bit error-prone having this logic in every function, maybe:
```python
return conditional_merge_datasets(ds, new_ds, merge)
```

File: sgkit/stats/hwe.py [hunk 2]
```diff
@@ -137,6 +139,12 @@ def hardy_weinberg_test(
         where `N` is equal to the number of variants and the 3 columns contain
         heterozygous, homozygous reference, and homozygous alternate counts
         (in that order) across all samples for a variant.
+    merge : bool, optional
+        If True (the default), merge the input dataset and the computed
```
Comment (jeromekelleher) [inline sgkit/stats/hwe.py:143]: I don't like the idea of duplicating this block of text in every function. Maybe we could refer to a section of the docs discussing what merge is, something like "If True (the default) merge the input dataset and the computed output variables into a single dataset. See :ref:`sec_dataset_merge` for more details".

Or, we could try to define a sphinx substitution or something?

**Review by jeromekelleher — APPROVED** [review]
Some minor comments.

File: sgkit/stats/association.py (hunk 1 above)
Comment (tomwhite) [reply sgkit/stats/association.py]: See https://github.com/pystatgen/sgkit/pull/250/commits/b634bd71dd70aaa3199399bf6e2c179c9eb4e6d2

File: sgkit/stats/hwe.py (hunk 2 above)
Comment (tomwhite) [reply sgkit/stats/hwe.py:143]: I prefer having a reference. Implemented in https://github.com/pystatgen/sgkit/pull/250/commits/cfc5e2f00d16d282e581ce6b707f08d43e161c63.

**Comment by jeromekelleher** [comment]
LGTM - @eric-czech, @ravwojdyla, it'd be good to get your eyes on this before we start duplicating the pattern.

**Review by eric-czech — APPROVED** [review]
LGTM

**Review by ravwojdyla — CHANGES_REQUESTED** [review]
@tomwhite could you please also update `pc_relate`?

**Comment by ravwojdyla** [comment]
@tomwhite thanks, `pc_relate` still has the old docstring, could you please update that as well?

**Comment by tomwhite** [comment]
Oops - should be fixed now.

**Review by ravwojdyla — APPROVED** [review]

**Pull request merged**